aggregator/src/compression/circuit.rs

```diff
@@ -0,0 +1,237 @@
+//! Circuit implementation for compression circuit.
+
+use std::fs::File;
+
+use ark_std::{end_timer, start_timer};
+use halo2_proofs::{
+    circuit::{Layouter, SimpleFloorPlanner, Value},
+    halo2curves::bn256::{Fq, G1Affine},
+    plonk::{Circuit, ConstraintSystem, Error},
+};
+use rand::Rng;
+use snark_verifier::{
+    loader::{
+        halo2::{
+            halo2_ecc::halo2_base::{
+                self,
+                halo2_proofs::{
+                    halo2curves::bn256::{Bn256, Fr},
+                    poly::{commitment::ParamsProver, kzg::commitment::ParamsKZG},
+                },
+                Context, ContextParams,
+            },
+            Halo2Loader,
+        },
+        native::NativeLoader,
+    },
+    pcs::kzg::{Bdfg21, Kzg, KzgAccumulator, KzgSuccinctVerifyingKey},
+    util::arithmetic::fe_to_limbs,
+};
+use snark_verifier_sdk::{aggregate, flatten_accumulator, types::Svk, Snark, SnarkWitness};
+
+use crate::{
+    core::extract_accumulators_and_proof,
+    param::{ConfigParams, BITS, LIMBS},
+};
+
+use super::config::CompressionConfig;
+
+/// Input a proof, this compression circuit generates a new proof that may have smaller size.
+///
+/// It re-exposes same public inputs from the input snark.
+/// All this circuit does is to reduce the proof size.
+#[derive(Clone)]
+pub struct CompressionCircuit {
+    pub(crate) svk: KzgSuccinctVerifyingKey<G1Affine>,
+    pub(crate) snark: SnarkWitness,
+    /// whether this circuit compresses a fresh snark
+    pub(crate) is_fresh: bool,
+    /// instances, flattened.
+    /// It re-exposes same public inputs from the input snark.
+    /// If the previous snark is already a compressed, this flattened_instances will
+    /// exclude the previous accumulator.
+    pub(crate) flattened_instances: Vec<Fr>,
+    // accumulation scheme proof, private input
+    pub(crate) as_proof: Value<Vec<u8>>,
+}
+
+impl Circuit<Fr> for CompressionCircuit {
+    type Config = CompressionConfig;
+    type FloorPlanner = SimpleFloorPlanner;
+
+    fn without_witnesses(&self) -> Self {
+        let flattened_instances = self
+            .snark
+            .instances
+            .iter()
+            .flat_map(|instance| instance.iter().map(|_| Fr::zero()))
+            .collect();
+
+        Self {
+            svk: self.svk,
+            snark: SnarkWitness::without_witnesses(&self.snark),
+            is_fresh: true,
+            flattened_instances,
+            as_proof: Value::unknown(),
+        }
+    }
+
+    fn configure(meta: &mut ConstraintSystem<Fr>) -> Self::Config {
+        // Too bad that configure function doesn't take additional input
+        // it would be nicer to load parameters from API rather than ENV
+        let path = std::env::var("VERIFY_CONFIG")
```

Comment:
suggest change name for this `VERIFY_CONFIG` to `COMPRESS_VERIFY_CONFIG`?  it is too general. eg: scroll-zkevm use this env var too.